ocaml/typing/jkind.mli

```diff
@@ -31,277 +31,422 @@
+      variable, it is set to [value] first. *)
+  val default_to_value_and_get : t -> Const.t
 
-  (* CR layouts: The [offender] arguments below are always
-     [Printtyp.type_expr], so we should either stash that in a ref (like with
-     [set_printtyp_path] below) or just move all the printing machinery
-     downstream of both [Jkinds] and [Printtyp]. *)
+  (** [default_to_value t] is [ignore (default_to_value_and_get t)] *)
+  val default_to_value : t -> unit
 
-  (* CR layouts: Having these options for printing a violation was a choice
-     made based on the needs of expedient debugging during development, but
-     probably should be rethought at some point. *)
+  (** [is_void t] is [Void = default_to_value_and_get t].  In particular, it will
+      default the jkind to value if needed to make this false. *)
+  val is_void_defaulting : t -> bool
+  (* CR layouts v5: When we have proper support for void, we'll want to change
+     these three functions to default to void - it's the most efficient thing
+     when we have a choice. *)
 
-  (** Prints a violation and the thing that had an unexpected jkind
-      ([offender], which you supply an arbitrary printer for). *)
-  val report_with_offender :
-    offender:(Format.formatter -> unit) -> Format.formatter -> t -> unit
+  (** Gets the layout of a jkind; returns [None] if the layout is still unknown.
+      Never does mutation. *)
+  val get_layout : t -> Layout.Const.t option
 
-  (** Like [report_with_offender], but additionally prints that the issue is
-      that a representable jkind was expected. *)
-  val report_with_offender_sort :
-    offender:(Format.formatter -> unit) -> Format.formatter -> t -> unit
+  (** Gets the maximum modes for types of this jkind. *)
+  val get_modal_upper_bounds : t -> Mode.Alloc.Const.t
 
-  (** Simpler version of [report_with_offender] for when the thing that had an
-      unexpected jkind is available as a string. *)
-  val report_with_name : name:string -> Format.formatter -> t -> unit
+  (** Gets the maximum mode on the externality axis for types of this jkind. *)
+  val get_externality_upper_bound : t -> Externality.t
+
+  (** Computes a jkind that is the same as the input but with an updated maximum
+      mode for the externality axis *)
+  val set_externality_upper_bound : t -> Externality.t -> t
+
+  (*********************************)
+  (* pretty printing *)
+
+  val format : Format.formatter -> t -> unit
+
+  (** Format the history of this jkind: what interactions it has had and why
+      it is the jkind that it is. Might be a no-op: see [display_histories]
+      in the implementation of the [Jkind] module.
+
+      The [intro] is something like "The jkind of t is". *)
+  val format_history :
+    intro:(Format.formatter -> unit) -> Format.formatter -> t -> unit
+
+  (*********************************)
+  (* debugging *)
+
+  module Debug_printers : sig
+    val t : Format.formatter -> t -> unit
+  end
+end
```

Comment:
Above this point, `module Type` exports existing interfaces, with some which are unused ones removed.  The visible diff includes the indentation.
Below, the interface for general `Jkind` is given.
I'd recommend using a diff viewer which does a better job with indentation than the one here.